### ISSUE TITLE
Reprint on current instrument with FX

### ISSUE THREAD

**Issue by tkna91** [esaruoho/org.lackluster.Paketti.xrnx#356]
I noticed today, isn't this mostly achieved with "Render To Sample"? In other words, it seems to me that there is nothing wrong with the standard feature itself.
It appears that this could be efficient with Paketti in the way that AFX has pointed out.
What do you think? 

Perhaps the following steps could be taken?

1. add a new pattern with any number of lines next to the current pattern
2. add C4 to the beginning of that pattern
3. Render Pattern to Sample that pattern
4. delete the pattern
5. delete the original Instrument
6. Change the newly created Instrument to the original Instrument number

Yeah, a simple test looks fine from an EQ perspective.

https://github.com/user-attachments/assets/128a494e-07bd-49e3-a5dc-bde56a889c0c

**Comment by tkna91** [comment]
I'm not sure, but isn't this it?
```lua
options = {
  start_pos,     -- renoise.SongPos object. by default the song start.
  end_pos,       -- renoise.SongPos object. by default the song end.
  sample_rate,   -- one of 22050, 44100, 48000, 88200, 96000, 192000. \
                 -- by default the players current rate.
  bit_depth ,    -- number, one of 16, 24 or 32. by default 32.
  interpolation, -- string, one of 'default', 'precise'. by default default'.
  priority,      -- string, one "low", "realtime", "high". \
                 -- by default "high".
}
```
> Parameter 'rendering_done_callback' is ONLY called when rendering has succeeded. You can do something with the file you've passed to the renderer here, like for example loading the file into a sample buffer.
```lua
renoise.song():render([options, ] filename, rendering_done_callback)
  -> [boolean, error_message or nil]
```

https://files.renoise.com/xrnx/documentation/Renoise.Song.API.lua.html#h2_36

**Comment by esaruoho** [comment]
api request please.
i can hack my way around it, somehow, by creating a new pattern, inputting notes to it, rendering it, and stuff, but it's not quite the same as having it as a api function.

i mean, it WILL work. but it'll be very twiddly.

**Comment by tkna91** [comment]
And it seems to me that if we use pattern-based rendering as in the above post, we can print destructively with reverb and delay samples without any problems.
This is assuming reverbs and delays, so it makes sense to implement.

**Comment by esaruoho** [comment]
yeah the tails will be there. won't be limited to "sample length = print with reverb = same sample length".

**Comment by tkna91** [comment]
Yes, the length of the sample to be rendered may be specified in the following way

- Length of the current pattern
- Specified by the number of lines by the user when rendering

It seems to me that it could be fixed at the length of the current pattern.

**Comment by esaruoho** [comment]
Am thinking exactly that yes. I hope tomorrow will be productive, I need to get below 100 tickets soon.

**Comment by tkna91** [comment]
It appears to be even better because it also has the following advantages

- The user does not have to consider whether the current sample is `TFX` or `SFX`.
- Can be used for samples using both `TFX` and `SFX

**Comment by esaruoho** [comment]
@tkna91 can you specify what the difference is between `SFX` and `TFX` - the terms are not immediately obvious to me.

is sfx sample fx chain and tfx "track fx" aka track dsp stuff?

**Comment by tkna91** [comment]
Yes, it is. It is the difference between track DSP and sample DSP.

![20240904-033142_screenshot](https://github.com/user-attachments/assets/ed971476-a052-4f61-a77e-804c649f851c)


![image](https://github.com/user-attachments/assets/5adad662-b3a2-45a4-b499-f010c29165d8)

**Comment by erroreyes** [comment]
_(Sorry if I haven't read the entire thread but I think I got the gist of it)_

Because it rubs me the wrong way to create a separate entry in a pattern in order to render a sample with effects and retain the tail, what I've done is create a tool that adds space at the end of a sample so that I can then render the sample with effect from the editor. The issue with that is that adding that extra space at the end of a sample takes a second to do, which is weird. It could be possible that I programmed the function wrongly, or it could be a limitation of the API, idk.

```lua
function add_silence()
    local buffer = renoise.song().selected_sample.sample_buffer
    local sr = buffer.sample_rate
    local bd = buffer.bit_depth
    local nc = buffer.number_of_channels
    local nf = buffer.number_of_frames
    -- len is the new length of the processed sample, which is the length of the original sample
    -- times some 'value' chosen by the user, for example selecting 4 from a dropdown menu
    local len = nf * value

    -- Temp table to hold the original sample data
    local tmpL = {}
    local tmpR = {}

    -- Save buffer to temp table
    for i = 1, nf do
        tmpL[i] = buffer:sample_data(LEFT, i)
        if nc == 2 then
            tmpR[i] = buffer:sample_data(RIGHT, i)
        end
    end

    -- Remove sample data from the buffer and set it with the orignal
    -- buffer's properties.
    buffer:create_sample_data(sr, bd, nc, len)

    -- Prepare the buffer for new sample data;
    -- Set sample data from the temp table to the current buffer;
    -- Finalize the buffer.
    buffer:prepare_sample_data_changes()

    for i = 1, len do
        if i <= nf then
            buffer:set_sample_data(LEFT, i, tmpL[i])
            if nc == 2 then buffer:set_sample_data(RIGHT, i, tmpR[i]) end                
        else
            buffer:set_sample_data(LEFT, i, 0)
            if nc == 2 then buffer:set_sample_data(RIGHT, i, 0) end 
        end
    end
    
    buffer:finalize_sample_data_changes()
end
```

![Peek 2024-09-11 15-33](https://github.com/user-attachments/assets/42dd3b59-1e56-440a-8608-f59a0ad8f915)

I also know that there is a tool shared somewhere in the forum that allows to do the 'add note -> render to new sample/instrument -> remove note' from one go, but I can't remember what it is called.

**Comment by esaruoho** [comment]
@erroreyes glad to have you thinking about this. okay, so, yes, your script adds silence to the end of the sample.
but the TFX or SFX button still needs to be pressed. my preference would be some sort of solution that does, with one shortcut, this:

1. adds silence to sample
2. renders effect to sample by pressing TFX or SFX

or

1. stores the current pattern
2. changes to a completely empty pattern
3. outputs C-4 note to that pattern and that track
4. renders the pattern length (let's say 512 rows max, to get the whole tail if possible)
5. deletes the empty pattern
6. returns back to original pattern
7. maybe even do a strip silence at 0.01% so when it is absolutely silent, cut the rest.

this would make it possible to do `TFX` without issues. since it would be that specific track DSP.
the SFX is actually simpler, all it would do is:

1. create a new track in the current pattern
2. select track
3. put C-4 of current instrument to track
4. render track
5. delete track
6. select the newly rendered sample/instrument

these would be pretty transparent ways of doing SFX and TFX rendering. without the user needing to do more than either press a shortcut, push a midi button, or select a menu entry.

it would not, however, overwrite the current sample, it would create a new sample.

i think this can be done and could be useful.
